app/Support/Shortcode/ExternalShortcode.php

```diff
@@ -0,0 +1,152 @@
+    {
+        $input = preg_replace('~\[img="?([^]"]*)"?]~i', '[img="$1"]', $input);
+        $input = preg_replace('~\[url="?([^]"]*)"?]~i', '[url="$1"]', $input);
+
+        foreach ($this->handlers->getNames() as $tag) {
+            $input = preg_replace("~\[/$tag]~i", "[/$tag]", $input); // closing tag
+            $input = preg_replace("~\[$tag]~i", "[$tag]", $input); // opening tag
+            $input = preg_replace("~\[$tag=~i", "[$tag=", $input); // opening tag with value
+        }
+
+        $events = new EventContainer();
+
+        $processor = (new Processor(new RegularParser(), $this->handlers))
+            ->withEventContainer($events);
+
+        $html = $processor->process($input);
+
+        // Remove any incomplete/truncated shortcodes.
+        // "Here is some content [url=https://ab..." -> "Here is some content ..."
+        if (!empty($html)) {
+            $html = preg_replace('~\s*\[/?\w+[^]]*$~', '...', $html);
+        }
+
+        return $html;
+    }
+
+    private function renderUrlLink(ShortcodeInterface $shortcode): string
+    {
+        $url = $shortcode->getBbCode() ?: $shortcode->getContent();
+        $prefixedUrl = $this->protocolPrefix($url);
+        $content = $shortcode->getContent() ?: $prefixedUrl;
+
+        return "<a target='_blank' rel='noreferrer' href='$prefixedUrl'>$content</a>";
+    }
+
+    private function renderLink(ShortcodeInterface $shortcode): string
+    {
+        $url = $shortcode->getParameter('url') ?: $shortcode->getContent();
+        $prefixedUrl = $this->protocolPrefix($url);
+        $content = $shortcode->getContent();
+
+        return "<a href='$prefixedUrl'>$content</a>";
+    }
+
+    private function renderSpoiler(ShortcodeInterface $shortcode): string
+    {
+        $content = $shortcode->getContent();
+
+        return "<span style='background: black; color: black;'>$content</span>";
+    }
+
+    private function protocolPrefix(?string $href): string
+    {
+        if (empty($href)) {
+            return '';
+        }
+
+        $scheme = parse_url($href, PHP_URL_SCHEME);
+        $host = parse_url($href, PHP_URL_HOST);
+
+        if (empty($scheme)) {
+            $href = 'https://' . ltrim($href, '/');
+        } elseif ($scheme === 'http' && str_ends_with($host, 'retroachievements.org')) {
+            $href = str_replace('http://', 'https://', $href);
+        }
+
+        return $href;
+    }
+
+    private function embedAchievement(int $id): string
+    {
+        $achievementData = Cache::store('array')->rememberForever('achievement:' . $id . ':card-data', fn () => GetAchievementData($id));
+
+        if (empty($achievementData)) {
+            return "<a href='/achievement/$id'>Achievement</a>";
+        }
+
+        $achievementName = $achievementData['Title'];
+        $achievementPoints = $achievementData['Points'];
+
+        return "<a href='/achievement/$id'>$achievementName ($achievementPoints)</a>";
```

Comment:
I can't help but notice that a large portion of this file is duplicated from Shortcode.php. Would it be possible to move the bits that need to be different into functions and override them in a subclass?